webpack.config.js

```diff
@@ -382,6 +379,112 @@ var config = {
     ],
 };
 
+async function initializeModuleFederation() {
+    function makeSingletonSharedModules(packageNames) {
+        const sharedObject = {};
+
+        for (const packageName of packageNames) {
+            const version = packageJson.dependencies[packageName];
+
+            sharedObject[packageName] = {
+                requiredVersion: version,
+                singleton: true,
+                version,
+            };
+        }
+
+        return sharedObject;
+    }
+
+    function isWebpackDevServerAvailable(baseUrl) {
+        return new Promise((resolve) => {
+            if (!DEV) {
+                resolve(false);
+                return;
+            }
+
+            const req = http.request(`${baseUrl}/remote_entry.js`, (response) => {
+                return resolve(response.statusCode === 200);
+            });
+
+            req.on('error', () => {
+                resolve(false);
+            });
+
+            req.end();
+        });
+    }
+
+    async function getRemoteModules() {
+        const products = [
+            {name: 'focalboard', baseUrl: 'http://localhost:9006'},
+        ];
+
+        const productsFound = await Promise.all(products.map((product) => isWebpackDevServerAvailable(product.baseUrl)));
+
+        const remotes = {};
+        const aliases = {};
+
+        for (let i = 0; i < products.length; i++) {
+            const product = products[i];
+            const found = productsFound[i];
+
+            if (found) {
+                console.log(`Product ${product.name} found, adding as remote module`);
+
+                remotes[product.name] = `${product.name}@${product.baseUrl}/remote_entry.js`;
+            } else {
+                console.log(`Product ${product.name} not found`);
+
+                // Add false aliases to prevent Webpack from trying to resolve the missing modules
+                aliases[product.name] = false;
+                aliases[`${product.name}/manifest`] = false;
+            }
+        }
+
+        return {remotes, aliases};
+    }
+
+    const {remotes, aliases} = await getRemoteModules();
+
+    config.plugins.push(new ModuleFederationPlugin({
```

Comment:
One more thing here, would it be safer to (since the first batch of PRs is about dev environments) to conditionally add the module federation plugin only for dev? I guess there is no impact since there won't be remotes in production, so the provided/fallback shared packages will be used for everything though.

Reply:
I think it should be fine to leave as-is for now. The module federation stuff will still be set up in production, but `isWebpackDevServerAvailable` always returns false as long as we're not using `npm run run` or `npm run dev-server`, so it shouldn't have any noticeable effect